### PR TITLE
fix: preserve FP8 KV tails in FA2 asymmetric prefill

### LINKED ISSUE #4976: [Bug] fa2 prefill with an fp8 KV cache is wrong for head_dim_qk=192 / head_dim_vo=128
### Before submitting

- [x] I have searched existing issues and did not find a solution or an existing report.

### Environment

GH200 (SM90), CUDA 12.8, torch 2.14.0+cu130, FlashInfer main at df8b5c17 (0.6.18).

### Bug description

`BatchPrefillWithRaggedKVCacheWrapper` with `backend="fa2"`, `head_dim_qk=192`, `head_dim_vo=128` (the DeepSeek prefill head-dim pair) and an fp8 KV cache returns wrong outputs for every row and every output dimension. The same call with a 16-bit KV cache is correct, and fp8 KV with `head_dim_qk == head_dim_vo` (64 / 128 / 256) is correct, so the defect is specific to the mixed-precision path with unequal head dims. Both `float8_e4m3fn` and `float8_e5m2`, both `float16` and `bfloat16` queries, causal and non-causal are affected.

Max abs error against an fp32 reference over the dequantized cache (99 tokens, 16 heads, randn inputs):

| query | KV cache | causal | max err |
|---|---|---|---|
| fp16 | e4m3 | no | 1.19 |
| fp16 | e5m2 | no | 1.49 |
| bf16 | e4m3 | no | 1.41 |
| bf16 | e5m2 | no | 1.34 |
| fp16 | e4m3 | yes | 2.49 |
| fp16 | fp16 | no | 0.0005 |

### Minimal reproduction

```python
import math, torch, flashinfer

torch.manual_seed(0)
seq_len, H, DQK, DVO = 99, 16, 192, 128
q_dtype, kv_dtype = torch.float16, torch.float8_e4m3fn
ws = torch.empty(128 << 20, dtype=torch.uint8, device="cuda")
q = torch.randn(seq_len, H, DQK, dtype=q_dtype, device="cuda")
k = torch.randn(seq_len, H, DQK, dtype=q_dtype, device="cuda").to(kv_dtype)
v = torch.randn(seq_len, H, DVO, dtype=q_dtype, device="cuda").to(kv_dtype)
indptr = torch.tensor([0, seq_len], dtype=torch.int32, device="cuda")

w = flashinfer.BatchPrefillWithRaggedKVCacheWrapper(ws, "NHD", backend="fa2")
w.plan(indptr, indptr, H, H, DQK, head_dim_vo=DVO, causal=False,
       q_data_type=q_dtype, kv_data_type=kv_dtype)
o = w.run(q, k, v).float()

s = torch.einsum("qhd,khd->hqk", q.float(), k.float()) / math.sqrt(DQK)
ref = torch.einsum("hqk,khd->qhd", torch.softmax(s, -1), v.float())
print("max err", (o - ref).abs().max().item())   # 1.19; 0.0005 with kv_dtype=torch.float16
```

### Additional context

Found while adding fa3 support for a 16-bit query with an fp8 KV cache (#3327); the fa3 kernel matches the reference on this head-dim pair, so the new test compares against fa3 rather than fa2 for it.

### PR DESCRIPTION
**Pull request by git-jxj**
## 📌 Description

FA2 prefill can return incorrect attention outputs with FP8 KV caches when a head dimension has a 64-byte tail, including the QK=192/VO=128 shape in #4976. The 128B load loop truncates the final block, and its XOR swizzle does not fit the row stride.

Use 64B KV swizzling for these FP8 shapes and load every 64B block in the contiguous, paged, and on-the-fly paged producers. Allow the existing 64B shared-memory helpers to address multiple blocks per row. This preserves the layout used by direct dequantization and the existing 16-bit repack path, without adding shared-memory padding.

Add reference-based regression coverage for single, ragged, and NHD/HND paged prefill, both FP8 formats, FP16/BF16 queries, causal/non-causal attention, short/long queries, and both asymmetric dimension directions.

## 🔍 Related Issues

Closes #4976.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

Validated from this branch with an isolated JIT cache on an NVIDIA RTX 5060 Ti (SM120, 16 GB), PyTorch 2.13.0+cu130, and CUDA 13.0. The numerical reference uses FP32 attention on the same quantized KV values; native FP16/BF16 KV runs are additional controls.

- Regression command: `python -m pytest -q tests/attention/test_fp8_prefill.py -k asymmetric -x`.
- Before the change, the QK=192/VO=128 ragged regression failed with maximum absolute error 0.9121 (94.1% of elements outside tolerance).
- The full regression matrix passed: **128 passed** (single/ragged/NHD/HND, two dimension directions, two query dtypes, two FP8 formats, two query lengths, causal/non-causal).
- Independent rerun of the original 99-token, 16-head example: maximum absolute error **0.0003118**, equal to the FP16 control, using **181.53 MiB** peak allocated GPU memory.
- `pre-commit run --all-files` passed on an isolated snapshot with the identical patch; changed-file hooks also passed.
- Supplemental controls passed: **24 existing FP8 calibration checks** (head dimensions 64/128/256, E4M3/E5M2, NHD/HND, query lengths 7/53) and **6 NVFP4 checks** using the existing test functions/helpers with additional head dimension 64 (single/ragged/paged, query lengths 7/64). The supplemental harness peaked at 296.23 MiB allocated GPU memory.

- After the dependent-template follow-up, **10 targeted GPU checks passed**: FP8 single/ragged/NHD/HND at query lengths 7/99 and FP4 paged prefill at head dimensions 64/128. A representative C++17 syntax check fails before adding the disambiguators and passes afterward; full `pre-commit run --all-files` also passed again.

## 🔬 Experimental Track

- [ ] This PR is **experimental**: it adds or changes code under `flashinfer/experimental/` and/or an `@flashinfer_experimental_api`. Tracking issue: #
  - [ ] The tracking issue names an owner, the reason for the experimental path, and a graduation plan with a target release.
  - [ ] Core changes are limited to a thin entry point (signature, shared validation, feature-gate check, backend selection, handoff).
  - [ ] Tests live in `tests/experimental/` and were validated on the intended hardware; a runnable example is included.
  - [ ] Nothing is registered in `flashinfer/aot.py`, and no experimental backend is reachable from `backend="auto"` without `FLASHINFER_ALLOW_EXPERIMENTAL_AUTO_BACKENDS=1`. (Calling an `@flashinfer_experimental_api` or naming a backend explicitly is itself the opt-in and needs no environment variable.)
  - [ ] **Test scope declared below.** The experimental CI lane runs exactly these targets, so keep them as narrow as the change allows.

```experimental-tests
# One target per line: a directory or a file. (A pytest ::selector is not
# supported -- the sharding runner cannot consume one.) Must be under
# tests/experimental/ and must exist. Delete these comment lines and add yours, e.g.
#
#   tests/experimental/test_my_backend.py
#   tests/experimental/my_backend/
#
# Declaring the whole tree (tests/experimental/) is allowed but means every
# experimental PR pays for every other feature's tests, in every matrix cell.
```

## Reviewer Notes

AI-assisted implementation and testing (Codex).

VO=192 coverage uses supported query tile sizes and disables split-KV for batch APIs. The unchanged base also rejects single QK=128/VO=192 with 99 query tokens on this GPU, and the batch merge kernel does not support VO=192, including with native FP16 KV. Those dispatch/merge limitations are outside this change; the original QK=192/VO=128 case retains 99 query tokens and default split-KV planning.

Validation is limited to SM120. `compute-sanitizer` is unavailable in the local environment.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved FP8 KV-cache prefill support for asymmetric query and value head dimensions, including dimensions with 64-byte tails.
  * Extended compatibility across single, ragged, NHD, and HND layouts, causal and non-causal modes, and supported FP8 and query data types.

* **Tests**
  * Added comprehensive coverage comparing FP8 prefill results with FP16 reference calculations.

### CODE REVIEW

File: include/flashinfer/attention/prefill.cuh [hunk 1]
```diff
@@ -618,16 +622,23 @@ __device__ __forceinline__ void produce_kv(smem_t<KTraits::SWIZZLE_MODE_KV> smem
...
-        smem.template load_64b_async<fill_mode>(*smem_offset, *gptr, kv_idx < kv_len);
-      } else {
-        smem.load_128b_async<fill_mode>(*smem_offset, *gptr, kv_idx < kv_len);
+#pragma unroll
+      for (uint32_t j = 0; j < NUM_MMA_D / (4 / sizeof(DTypeKV)); ++j) {
+        // FP4 GMEM rows are packed 2x denser; load 64b (upper 64b of smem slot zeroed).
+        if constexpr (IS_FP4) {
+          smem.template load_64b_async<fill_mode>(*smem_offset, *gptr, kv_idx < kv_len);
+        } else {
+          smem.load_128b_async<fill_mode>(*smem_offset, *gptr, kv_idx < kv_len);
```
[the remark this answers is not in the document]
Comment (git-jxj) [reply include/flashinfer/attention/prefill.cuh]: Fixed in a27c1d7. Added the missing `template` disambiguators to the three FP8 calls and the two adjacent FP4 calls. The C++ syntax check, 10 targeted GPU checks, and all pre-commit checks passed.

**Comment by git-jxj** [comment]
CI for current head `a27c1d78efaf` is still pending authorization: [Test Results Summary](https://github.com/flashinfer-ai/flashinfer/actions/runs/34453963847/job/102795936294) reports `CI skipped — pending authorization`, and the test matrix was skipped. Could a maintainer or `ci-users` member start validation of this head with `@flashinfer-bot run`?
